### ISSUE TITLE
Could not find value for entry assembly and version type Assembly - Only when async=true

### ISSUE THREAD

**Issue by Giorgi** [NLog/NLog.Web#633]
I get `Could not find value for entry assembly and version type Assembly` but only if `async="true"` is specified.

**Type** (choose one):

- Bug

**NLog version**: 4.7.7

**NLog.Web version**: 4.10.0

**NLog.Extensions.Logging version**: 

**Platform**:  .Net 4.6.1

**Current NLog config** (xml or C#, if relevant)

```xml
<?xml version="1.0" encoding="utf-8" ?>
<nlog xmlns="http://www.nlog-project.org/schemas/NLog.xsd"
      xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">

  <targets async="true">
    <target name="db" xsi:type="Database" connectionStringName="PRODUCTION">
      <commandText>
        INSERT INTO [dbo].[ApplicationLog] ([LogLevel], [Logger], [UserId], [Message], [ErrorMessage], [ProjectName], [Url],
        [ServerIp], [ServerName],[RequestIp], [ProcessId], [ThreadId], [Properties], [AppVersion], [RequestId])
        VALUES (@LogLevel, @Logger, @UserId, @Message, @ErrorMessage, @ProjectName, @Url, @ServerIp, @ServerName, @RequestIp, @ProcessId, @ThreadId, @Properties, @AppVersion, @RequestId)</commandText>
      <parameter name="@LogLevel" layout="${level}" dbType="SqlDbType.Varchar" />
      <parameter name="@Logger" layout="${logger}" dbType="SqlDbType.Varchar" />
      <parameter name="@UserId" layout="${mdlc:item=UserId}" dbType="SqlDbType.Varchar"  />
      <parameter name="@Message" layout="${message}" dbType="SqlDbType.VarChar" />
      <parameter name="@ErrorMessage" layout="${exception:format=ToString:maxInnerExceptionLevel=5:innerFormat=ToString}" dbType="SqlDbType.VarChar"  />
      <parameter name="@ProjectName" layout="${gdc:item=Project}" dbType="SqlDbType.VarChar" />
      <parameter name="@Url" layout="${aspnet-request-url:IncludeQueryString=true}" dbType="SqlDbType.VarChar"  />
      <parameter name="@ServerIp" layout="${local-ip:cachedSeconds=60}" dbType="SqlDbType.VarChar"  />
      <parameter name="@ServerName" layout="${machinename}" dbType="SqlDbType.VarChar" />
      <parameter name="@RequestIp" layout="${aspnet-request-ip:CheckForwardedForHeader=true}" dbType="SqlDbType.Varchar" />
      <parameter name="@ProcessId" layout="${processid}" dbType="SqlDbType.Int"/>
      <parameter name="@ThreadId" layout="${threadid}" dbType="SqlDbType.Int"/>
      <parameter name="@Properties">
        <layout type="JsonLayout" includeAllProperties="true" maxRecursionLimit="1" escapeForwardSlash="false" />
      </parameter>
      <parameter name="@AppVersion" layout="${assembly-version}" dbType="SqlDbType.VarChar" />
      <parameter name="@RequestId" layout="${mdlc:item=RequestId}" dbType="SqlDbType.VarChar" />
    </target>
  </targets>

  <rules>
    <logger name="*" minlevel="Debug" writeTo="db" />
  </rules>
</nlog>
```

In case of a BUG:

- What is the current result? - Could not find value for entry assembly and version type Assembly
- What is the expected result? - assembly version is retrieved and logged
- Did you checked the [Internal log](https://github.com/NLog/NLog/wiki/Internal-Logging)? - Yes
```
2021-02-16 23:16:12.8866 Trace Extending ${assembly-version} AssemblyVersionLayoutRenderer with NLog.Web implementation
2021-02-16 23:16:12.9066 Trace   DatabaseTarget: Parameter: '@AppVersion' = 'Could not find value for entry assembly and version type Assembly' (AnsiString)
2021-02-16 23:16:12.9172 Trace   DatabaseTarget: Parameter: '@RequestId' = '00d9b2dd-98c8-48d4-92d1-678d3ccd20bc' (AnsiString)
2021-02-16 23:16:12.9172 Error DatabaseTarget(Name=db_wrapped): Error when writing to database. Exception: System.Data.SqlClient.SqlException (0x80131904): String or binary data would be truncated.
The statement has been terminated.
```
- Are there any workarounds? yes/no - Removing async "solves" the issue
- Is there a version in which it did work?  - No
- Can you help us by writing an unit test? - Probably not

**Comment by snakefoot** [comment]
Thank you for reporting this bug.

NLog.Web has a special `AssemblyVersionLayoutRenderer` to resolve this issue: https://github.com/NLog/NLog.Web/issues/46

https://github.com/NLog/NLog.Web/blob/master/src/Shared/LayoutRenderers/AssemblyVersionLayoutRenderer.cs

But I broke it with this https://github.com/NLog/NLog.Web/pull/326 where I added these class-attributes:
```c#
    [ThreadAgnostic]
    [ThreadSafe]
```

But `[ThreadAgnostic]` should not have been added, since the NLog.Web-version depends on the active HttpContext to work. When using `<targets async="true">` then the HttpContext is no longer available, and so it fails.

**Comment by Giorgi** [comment]
Yes, it loads the renderer from NLog.Web as I saw in the internal log: `Extending ${assembly-version} AssemblyVersionLayoutRenderer with NLog.Web implementation`

**Comment by snakefoot** [comment]
Created #634 to revert the change, but will have to make an unit-test aswell. To ensure the class-attributes are handled correctly.

**Comment by snakefoot** [comment]
Seems the bug was initially introduced with #305 (but it did not become better with #326)

But #634 needs more work, since it is not enough to just remove `[ThreadAgnostic]`

**Comment by snakefoot** [comment]
Have now updated #634 with a proper fix, and an unit test to prove it is working.

**Comment by Giorgi** [comment]
Also, I think that the 

`Could not find value for entry assembly and version type Assembly`

message should get logged to internal log instead of the target log destination.

**Comment by snakefoot** [comment]
> message should get logged to internal log instead of the target log destination.

That sounds like a different issue. You are welcome to create a pull-request for the NLog-project to improve this.

**Comment by snakefoot** [comment]
Bug has been resolved by #634

**Comment by Giorgi** [comment]
> > message should get logged to internal log instead of the target log destination.
> 
> That sounds like a different issue. You are welcome to create a pull-request for the NLog-project to improve this.

Agreed

**Comment by Giorgi** [comment]
Is there any target date for 4.10.1 milestone?

**Comment by snakefoot** [comment]
@Giorgi I think it would make sense to first release [NLog.Extensions.Logging ver. 1.7.1](https://github.com/NLog/NLog.Extensions.Logging) and then upgrade this repository and then release ver. 4.10.1 (Make sure that NLog.Web and NLog.Web.AspNetCore both depends on NLog 4.7.7)

But it depends on how the time schedule looks for @304NotModified ?

**Comment by 304NotModified** [comment]
Will work on the releases this weekend, or if I find time, earlier :)

**Comment by 304NotModified** [comment]
@Giorgi it's released as version 4.11 and now live: https://www.nuget.org/packages/NLog.Web.AspNetCore/4.11.0
